Style guide (see #3) (#42)
fortran-lang/stdlib: STYLE_GUIDE.md

# Fortran stdlib Style Guide

Adopting a consistent style can improve code legibility through the choice of good naming conventions.
In addition, style checks will be run during CI to flag any severe non-conformance.
This allows code review discussions to focus on semantics and substance rather than pedantry.

> Consistent whitespace usage, and not poluting line endigs with trailing white space makes `git diff`s considerably more legible.

Reviewer: ```suggestion
Consistent whitespace usage, and not polluting line endings with trailing white space makes `git diff`s considerably more legible.
```

Author: Speaking of typos, can I add https://github.com/reviewdog/action-misspell as a github action? It may need a new/different token setup. It creates really nice PR comments and would have caught this.

# Fortran stdlib Style Guide

> [A]utomate and codify as much as you possibly can while remembering that the human touch is still necessary to praise
> and be kind to your contributors.
> Let robots handle your project’s pedantry and humans handle your project’s empathy.

-- @mikemcquaid, [Homebrew] project leader<sup>[1]</sup>

[1]: https://mikemcquaid.com/2018/06/05/robot-pedantry-human-empathy/

> [Homebrew]: https://brew.sh

Reviewer 2: Personally not crazy about this quote.

Reviewer: Yeah, me neither.

I do want to automate formatting and everything else. But the quote doesn't quite get get excited.

Author: Haha, OK, that's fine with me. The point was just to convey a need for human empathy, and to highlight the fact to other maintainers that our time is better spent on substantive discussions. Where people have strong opinions about, e.g., style, then it's nice to have a robot complaining about it instead of a maintainer. Contributors tend to not take offense or want to argue, and everyone else gets to spend more time doing productive work. If you want to remove it, that's fine by me. BTW, I would recommend reading the article linked, and also another one about maintaining open source and the user-contributor-maintainer funnel (like a sales funnel).

Author: Here is the contributor funnel article, if you're interested: https://mikemcquaid.com/2018/08/14/the-open-source-contributor-funnel-why-people-dont-contribute-to-your-open-source-project/

And here is the article that the quote is from: https://mikemcquaid.com/2018/06/05/robot-pedantry-human-empathy/

I think both are definitely worth a read.

## File naming conventions

> - Source files should contain at most one `program` or `module`/`submodule`.

Reviewer 2: ```suggestion
- Source files should contain at most one `program`, `module`, or `submodule`.
```

> - The filename should match the program or module name and have the file extension `.f90` or `.F90` if preprocessing is required.

Reviewer 2: As far as I understand, `.F90` for files to be preprocessed is only useful for the compiler to imply preprocessing. We can explicitly enable preprocessing with a flag (`-cpp`). Should we recommend just the `.f90` suffix?

Author: There is a *defacto* convention for using `.F90` and letting the compiler do the preprocessing. I would rather stick to this established convention. It indicates to the reader what to expect and works more smoothly with compilers and build tools like CMake.

- If the interface and implementation is split using submodules the implementation submodule file should have the same name as the interface (parent) module but end in `_implementation`. E.g., `string_class.f90` and `string_class_implementation.f90`
- Tests should be added in the `tests` subdirectory and have the same name as the module they are testing with the `test_` prefix added. E.g., `string_class.f90` and `tests/test_string_class.f90`

## Indentation & whitespace

By setting and following a convention for indentation and whitespace, code reviews and git-diffs can
focus on the semantics of the proposed changes rather than style and formatting.

* __TABs vs spaces__: __Spaces__: TABs should never be used for indentation
* __Trailing whitespace__: All trailing whitespace must be removed
* __End of File (EOF)__: The file must end in a single new-line character

> * __Line length__: __132__: The Fortran 90 (and later) standard mandates lines be no longer than 132 characters

Reviewer 2: This whole section seems overly technical, rather than part of the guide. I suggest we extract the actionable suggestions and list them in plain language, for example:

* The body of every Fortran construct should be indented by 4 spaces
* Line length should be limited to 80 characters, but must not exceed 132.

If trailing whitespaces and EOL and EOF stuff will be handled by CI, no need to mention it in the guide.

Reviewer: Regarding 80 characters, I follow very strictly 80 character limit (and frequently have several terminals side by side on a screen, which the 80 characters allow), but not everybody does, and so I like the way you phrased it as `Line length should be limited to 80 characters and must not exceed 132.`, to recommend to try to keep it under 80, but if it can't be done, then one can go over. @zbeekman is that an acceptable compromise for you?

Author: @certik yes, I like the way I phrased the line length stuff to address line length. Keeping it less than 80 is ideal, but under 132---to provide an absolute limit somewhere, and conform to the standard---is a good compromise IMO. Especially if you have 4 spaces of indentation.

@milancurcic : Well, when you say:

> If trailing whitespaces and EOL and EOF stuff will be handled by CI, no need to mention it in the guide.

what are you envisioning?

If contributors aren't notified ahead of time about CI checks it could be frustrating to learn after you've committed code that you need to make stylistic changes. Or are you suggesting that we have CI automatically alter and commit changes to contributed code so that it would conform with the style?

Author: I've done my best to simplify and cleanup this section using more plain language, and respond to your suggestions.

@milancurcic I've included some stuff about tabs and trailing white space because I think it's worthwhile to be up-front about it, and don't envision a CI system commiting changes on a users behalf.

Reviewer 2: > Or are you suggesting that we have CI automatically alter and commit changes to contributed code so that it would conform with the style?

Sorry, indeed I assumed this would be done automatically via CI. I agree, if the developer is to worry about these things, we should state it in the style guide.

Author: We can setup pre-commit hooks to automatically fix issues like this and do auto-formatting, but it's hard to ensure that people are always using them.

In CI, I don't think it makes sense (at least not right now) to have CI authoring commits on the users behalf.

* __Indentation__: __Two (2) spaces__ for all control constructs, line continuations, etc.
* __Line ending character__: __Native__: Git will convert line endings to `\n` (Unix style) on commit, and will checkout files with native line endings

> ## Variable naming

Reviewer 2: ```suggestion
## Variable and procedure naming
```

Variable names should be descriptive, and not artificially short.
By default, where it makes sense to do so, variable names shall be made up of one or more full words separated by an underscore.

> For cases where a conventional & appropriate shortening of a word is used then the underscore may be omitted.

Reviewer: I suggest to rework this section. 

Here is a suggestion (from https://www.fortran90.org/src/best-practices.html#naming-convention):

---

* Use lowercase for all Fortran constructs (`do`, `subroutine`, `module`, ...).
* Follow short mathematical notation for mathematical variables/functions (`Ylm`, `Gamma`, `gamma`, `Enl`, `Rnl`, ...).
* For other names use all lowercase: try to keep names to one or two syllables; if more are required, use underscores to clarify (`sortpair`, `whitechar`, `meshexp`, `numstrings`, `linspace`, `meshgrid`, `argsort`, `spline`, `spline_interp`, `spline_interpolate`, `stoperr`, `stop_error`, `meshexp_der`).

For example "spline interpolation" can be shortened to `spline_interpolation`, `spline_interpolate`, `spline_interp`, `spline`, but not to `splineint` ("int" could mean integration, integer, etc. --- too much ambiguity, even in the clear context of a computational code). This is in contrast to `get_argument()` where `getarg()` is perfectly clean and clear.

The above are general guidelines. In general, choosing the right name certainly depends on the word being truncated as to whether the first syllable is sufficient. Usually it is but clearly not always. Thus some thought should go into step "try to keep names to 2 syllables or less" since it can really affect the indicativeness and simplicity. Simple consistent naming rules are a real help in this regard – for both collaboration and for one’s own sanity when going back to some old code you haven’t seen in while.

---

Here is the source code so that you can copy & paste it:
```markdown
* Use lowercase for all Fortran constructs (`do`, `subroutine`, `module`, ...).
* Follow short mathematical notation for mathematical variables/functions (`Ylm`, `Gamma`, `gamma`, `Enl`, `Rnl`, ...).
* For other names use all lowercase: try to keep names to one or two syllables; if more are required, use underscores to clarify (`sortpair`, `whitechar`, `meshexp`, `numstrings`, `linspace`, `meshgrid`, `argsort`, `spline`, `spline_interp`, `spline_interpolate`, `stoperr`, `stop_error`, `meshexp_der`).

For example "spline interpolation" can be shortened to `spline_interpolation`, `spline_interpolate`, `spline_interp`, `spline`, but not to `splineint` ("int" could mean integration, integer, etc. --- too much ambiguity, even in the clear context of a computational code). This is in contrast to `get_argument()` where `getarg()` is perfectly clean and clear.

The above are general guidelines. In general, choosing the right name certainly depends on the word being truncated as to whether the first syllable is sufficient. Usually it is but clearly not always. Thus some thought should go into step "try to keep names to 2 syllables or less" since it can really affect the indicativeness and simplicity. Simple consistent naming rules are a real help in this regard – for both collaboration and for one’s own sanity when going back to some old code you haven’t seen in while.
```

Author: I followed @milancurcic's suggestions from below but I'm happy to iterate on this more too. I'm marking as resolved because its outdated and changed, but if you feel strongly that what is there now needs changing, please don't hesitate to say so.

Examples:

__GOOD__:

``` fortran

> logical :: has_failed

Reviewer: Based on my above comment, both `has_failed` and `hasfailed` can be used, depending on the situation.

real function linspace(...)
```

__BAD__:

``` fortran
logical :: has_failed
real function lin_space(...)

> ```

Reviewer 2: (I can't seem to work out GitHub suggestion for multi-lines)

This whole thing can be condensed to:

* Variable names should be made up of one or more full words separated by an underscore, for example `has_failed` is preferred over `hasfailed`
* Where conventional and appropriate shortening of a word is used then the underscore may be omitted, for example `linspace` is preferred over `lin_space`

Reviewer: I think `has_failed` and `lin_space` is equivalent. They could both be used, but if it's just two syllables, it is perfectly fine to remove the `_` and just make it `linspace` and `hasfailed`. It all depends on the context if one should be preferred over the other.

In general, it's worth not to be typing `_` all the time, and it actually somtimes improves readability to remove `_` for two syllable names. If it's more than two syllables, then the `_` improves readability.

Reviewer 2: I see your point but I don't think they're quite equivalent. I see `linspace` is a programming portmanteau for "linear space". `linspace` makes more sense because `lin` is not a word on its own. `has_failed` however are two independent words, and like `is`, word `has` implies a logical state. I find `hasfailed` about as readable as the other one, but `has_*` has a semantic meaning implied in it (similar to `is_`, `get_`, `set_`, etc.).

Perhaps these are not the clearest example for the guide because the difference is nuanced.

Reviewer: Yes, I agree that `is_alpha` as we discussed should be with `_` because `is` has a semantic meaning. In the same spirit, if that is the case with `has`, then it should be `has_failed`. (All I was saying is that I can imagine a context when `has` does not have a semantic meaning, in which case `hasfailed` works also.) So I wanted to keep our rules general enough to allow for both, based on the context.

Author: As a starting point I'm adopting @milancurcic's text/suggestion, but I'm happy to iterate on it more if someone can propose a better alternative that we agree upon.

> ## Keyword case

Reviewer 2: This section can be dropped if we recommend lowercase for all variable and procedure names, as well as Fortran keywords, as in suggestion above.

Fortran keywords should __*not*__ be UPPERCASE.

> Modern editors and IDEs can highlight Fortran syntax and UPPERCASE keywords.

Reviewer: What do you mean by the sentence `Modern IDEs can highlight UPPERCASE keywords`? Are you trying to say `Modern IDEs can make lowercase keywords uppercase just for the screen (they stay lowercase in the file)`? Or are you trying to say `Modern IDEs can only highlight UPPERCASE keywords`? Or is it a typo and you meant `Modern IDEs can highlight lowercase keywords`.

Author: I think I should get rid of that `UPPERCASE`, it's not pertinent.

Author: This was obsoleted and removed due to other suggestions, marking resolved.

UPPERCASE keywords give Fortran source code the appearance of being antiquated.

## End <scope> block closing statements

Fortran allows certain block constructs or scopes to include the name of the program unit in the end statement.

> The convention adopted herein is to __NOT__ include procedure names, `module` names and `program` names in the `end` statement.

Reviewer 2: I disagree here. stating the name of the unit on `end` statements help readers see what exactly is being ended.

Reviewer: I do it sometimes for longer subroutines. For shorter ones I usually don't, so I quite liked the suggestion. But we can remove it, not a big deal.

Author: Yeah, I started the PR mostly as a place to discuss particulars. I'm not hugely against adding the ending names other than the fact that for short blocks it's more typing in an already chatty language.

Author: Also, I should note, there are only so many things that can be programmatically enforced for Fortran; its grammar is fairly complex and there are only so many tools that can enforce it. Fortunately, I think *requiring*/*adding* matching named end constructs is doable with `findent`. My suggestion as it stands is not, as far as I can tell, enforceable.

## File naming conventions

* Source files should contain at most one `program`, `module`, or `submodule`

> * The filename should match the program or module name and have the file extension `.f90` or `.F90` if preprocessing is required

Reviewer 3: The first bullet and last bullet points are more or less the same.

Author: Good point. I was moving this from another section and didn't realize it was already covered here. Thanks!
